### PR TITLE
feat(compiler): add support for `string` on `+=` for assignaments

### PR DESCRIPTION
**Pull request by Warkanlock**
Related to #5383

## Description

Based on #5383 request, we wanted to add the `+=` operator support to `string` types, so we are now able to achieve the following and compile `.wing` files correctly

```rust
let var initial = "hey";
initial += "you!";
```

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: examples/tests/invalid/types_strings_arithmetic.test.w [hunk 1]
```diff
@@ -8,4 +8,9 @@ let strExample = "Hello World";
 let numExample = 2;
 
 let e3 = "{strExample}!" * numExample;
-// ERR   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Cannot multiply a string with a number
\ No newline at end of file
+// ERR   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Cannot multiply a string with a number
+
+let a = 2;
+let b = "2";
+let b += a;
```
Comment (yoav-steinberg) [inline examples/tests/invalid/types_strings_arithmetic.test.w:15]: You probably wanted the third assignment to update `b` and note re-declare it.
```suggestion
let a = 2;
let var b = "2";
b += a;
```

File: examples/tests/valid/std_string.test.w [hunk 2]
```diff
@@ -18,6 +18,10 @@ assert("hello" + " world" == "hello world");
 assert(
 "hello {"funky"}
  world" == "hello funky\n world");
+
+let var initial = "hey, ";
+initial += "you";
+assert(initial == "hey, you");
```
Comment (yoav-steinberg) [inline examples/tests/valid/std_string.test.w:24]: :clap:

File: libs/wingc/src/lib.rs [hunk 3]
```diff
@@ -276,6 +276,7 @@ pub fn compile(
 	// Wing files, then move on to files that depend on those, and repeat)
 	for file in &topo_sorted_files {
 		let mut scope = asts.remove(file).expect("matching AST not found");
+
```
Comment (yoav-steinberg) [inline libs/wingc/src/lib.rs:279]:
```suggestion
```
Lets keep the diff with only relevant changes

File: libs/wingc/src/parser.rs [hunk 4]
```diff
@@ -1924,6 +1924,7 @@ impl<'s> Parser<'s> {
 	fn build_reference(&self, reference_node: &Node, phase: Phase) -> DiagnosticResult<Expr> {
 		let actual_node = reference_node.named_child(0).unwrap();
 		let actual_node_span = self.node_span(&actual_node);
+
```
Comment (yoav-steinberg) [inline libs/wingc/src/parser.rs:1927]:
```suggestion
```

File: libs/wingc/src/parser.rs [hunk 5]
```diff
@@ -1983,6 +1984,7 @@ impl<'s> Parser<'s> {
 		let expression_span = self.node_span(exp_node);
 		CompilationContext::set(CompilationPhase::Parsing, &expression_span);
 		let expression_node = &self.check_error(*exp_node, "expression")?;
+
```
Comment (yoav-steinberg) [inline libs/wingc/src/parser.rs:1987]:
```suggestion
```

File: libs/wingc/src/type_check.rs [hunk 6]
```diff
@@ -4247,12 +4248,22 @@ impl<'a> TypeChecker<'a> {
...
+			}
+			_ if var.type_.is_subtype_of(&self.types.number())
+				&& matches!(kind, AssignmentKind::AssignIncr | AssignmentKind::AssignDecr) =>
+			{
+				self.validate_type(exp_type, self.types.number(), value);
+				self.validate_type(var.type_, self.types.number(), variable);
+			}
+			_ => {
+				self.validate_type(exp_type, var.type_, value);
+			}
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:4265]: The logic seems right, but using `match` here when everything is prefixed with a `_` seems cumbersome.
Try changing to a simple `if .. else if ..` chain:
```rust
if matches!(kind, AssignmentKind::AssignIncr) {
  self.validate_type_in(expr_type, &[self.types.number(), self.types.string()], value);
  self.validate_type_in(var.type_, &[self.types.number(), self.types.string()], value);
} else if matches!(kind, AssignmentKind::AssignDecr) {
  self.validate_type(expr_type, self.types.number(), value);
  self.validate_type(var.type_, self.types.number(), value);
}
```
or something similar

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 7]
```diff
@@ -3659,7 +3659,21 @@ Duration <DURATION>"
...
+
+
+error: Unexpected 'identifier'
+   --> ../../../examples/tests/invalid/types_strings_arithmetic.test.w:15:10
+   |
+15 | let b += a;
+   |          ^ Unexpected 'identifier'
+
+
+error: Binary operator '+' cannot be applied to operands of type 'num' and 'str'; only (num, num) and (str, str) are supported
```
Comment (yoav-steinberg) [inline tools/hangar/__snapshots__/invalid.ts.snap:3676]: this is a good hint that the test isn't doing what we expected it to do

File: libs/wingc/src/type_check.rs (hunk 6 above)
Comment (Warkanlock) [reply libs/wingc/src/type_check.rs:4265]: I was on the same line before making this change (second commit); you are right, I'll rollback

Comment (Warkanlock) [reply libs/wingc/src/type_check.rs:4265]: In addition, thanks for the validate_type_in; I didn't know it's existence

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 7 above)
Comment (Warkanlock) [reply tools/hangar/__snapshots__/invalid.ts.snap:3676]: Is there an easy way to generate this `hangar/*.snap` files from the CLI? 

is it ``pnpm turbo test`` sufficient?

Comment (yoav-steinberg) [reply tools/hangar/__snapshots__/invalid.ts.snap:3676]:
```
cd tools/hangar
turbo test
```
But that still takes a pretty long time. There's probably a quicker way, but I'm not sure, try the dev channel on slack.
You can also just run the compiler on the specific test and see the output:
```
cd libs/wingc
cargo run --example compile -- ../../examples/tests/invalid/something.test.w
```

Comment (Warkanlock) [reply tools/hangar/__snapshots__/invalid.ts.snap]: thanks!

**Review by yoav-steinberg — APPROVED** [review]
:partying_face:

**Comment by NathanTarbert** [comment]
Thanks, @Warkanlock for your contribution! ❤️

**Pull request merged**